kwasm-lemmas.md

```diff
@@ -13,24 +13,28 @@ Basic arithmetic
 ----------------
 
 ```k
-    rule (0 +Int X) => X
-    rule (X +Int 0) => X
+    rule (0 +Int X) => X [simplification]
+    rule (X +Int 0) => X [simplification]
 
     rule (X +Int (Y *Int N)) modInt N => X modInt N
       requires N >=Int 1
```

Comment:
Can this be generalized to `N =/=Int 0` instead? It then looks more similar to the rule for `/Int` below.

Though, not sure what `modInt` with a negative number means...

Reply:
Yeah, usually in modular arithmetic the modulus is assumed to be a positive integer, and different languages implement the negative case differently. In Python `12 % -5 == -3`, in Java `12 % -5 == 2`, and in Haskell
```
12 `mod` (-5) == -3
12 `rem (-5) == 2
```
Either way, this should hold, but I don't want to get too general due to the weirdness of mod/rem.

Reply:
Finally, in z3 the result of mod is always non-negative, it seems. The following is unsat:

```z3
(declare-const x   Int)
(declare-const y   Int)
(declare-const res Int)

(assert (< y 0))
(assert (< res 0 ))
(assert (= res (mod x y)))
```

Reply:
We could of course make the condition `N =/=Int 0 andBool notBool (N <Int 0)` or something, if that will make resoning easier for the prover.

Reply:
And for the sake of establishing trust in the trusted codebase, this verifies the claim in z3, starting from the assumption that `mod` distributes.

```
(declare-const x   Int)
(declare-const y   Int)
(declare-const n   Int)

(assert (forall ((i Int) (j Int) (k Int)) (= (mod (+ i j) k) (+ (mod i k) (mod j k))))) ; mod distribtues over addition.

(assert (< 0 n))
(assert (not (= (mod (+ x (* y n)) n) (mod x n))))

(check-sat)
```
Gives `unsat`.